backend/core/src/shard/connector.rs

```diff
@@ -0,0 +1,156 @@
+use std::mem;
+use std::time::{Duration, Instant};
+
+use crate::aggregator::{AddNode, Aggregator};
+use crate::chain::{Chain, RemoveNode, UpdateNode};
+use crate::shard::ShardMessage;
+use crate::types::NodeId;
+use crate::util::{DenseMap, Hash};
+use actix::prelude::*;
+use actix_http::ws::Item;
+use actix_web_actors::ws::{self, CloseReason};
+use bincode::Options;
+use bytes::{Bytes, BytesMut};
+
+/// How often heartbeat pings are sent
+const HEARTBEAT_INTERVAL: Duration = Duration::from_secs(20);
+/// How long before lack of client response causes a timeout
+const CLIENT_TIMEOUT: Duration = Duration::from_secs(60);
+/// Continuation buffer limit, 10mb
+const CONT_BUF_LIMIT: usize = 10 * 1024 * 1024;
+
+pub struct ShardConnector {
+    /// Client must send ping at least once every 60 seconds (CLIENT_TIMEOUT),
+    hb: Instant,
+    /// Aggregator actor address
+    aggregator: Addr<Aggregator>,
+    /// Genesis hash of the chain this connection will be submitting data for
+    genesis_hash: Hash,
+    /// Chain address to which this multiplex connector is delegating messages
+    chain: Option<Addr<Chain>>,
+    /// Mapping `ShardConnId` to `NodeId`
+    nodes: DenseMap<NodeId>,
+    /// Buffer for constructing continuation messages
+    contbuf: BytesMut,
+}
+
+impl Actor for ShardConnector {
```

Comment:
Is there an `Actor` trait coming from `use actix::prelude::*` for this impl.

Reply:
Indeed, it's [here](https://docs.rs/actix/0.11.1/actix/trait.Actor.html).

Notice how the trait provides default implementations for most (all?) methods, which means that people like us only has to implement whatever methods we need to override and let the defaults be.

Reply:
Hence, the `fn started`, and `fn stopped` override. That's really useful. That makes a lot of sense now for when I am looking at any `impl` and considering what traits are being imported!